### ISSUE TITLE
how long does it take to train?

### ISSUE THREAD

**Issue by xchgaur** [daniel-kukiela/nmt-chatbot#44]
I am running two instances of nmt-chatbot in two separate setups. I have taken two deep learning VMs from azure: NC24 & NC12  with 24 & 12 vCPUs respectively.
In one instance I have used the dat as it is from you git repo. In the other instance I have created training data myself for reddit as instructed by Sentdex.  The training program  has been running on both the VMs since last 5 days.
Is there an indicator which can tell me that the training is going to complete?
Below is the  latest status on both the VMs:

# default data VM1

  global step 178400 lr 0.001 step-time 2.46s wps 2.36K ppl 4.52 gN 16.79 bleu 1.08
  global step 178500 lr 0.001 step-time 2.32s wps 2.58K ppl 4.46 gN 15.53 bleu 1.08
  global step 178600 lr 0.001 step-time 2.31s wps 2.59K ppl 4.51 gN 15.54 bleu 1.08
  global step 178700 lr 0.001 step-time 2.31s wps 2.59K ppl 4.64 gN 16.18 bleu 1.08
  global step 178800 lr 0.001 step-time 2.30s wps 2.60K ppl 4.75 gN 16.67 bleu 1.08
  global step 178900 lr 0.001 step-time 2.32s wps 2.60K ppl 4.91 gN 16.41 bleu 1.08

#VM2: with larger training data

	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,?], [?,?], [?,?], [?], [?]], output_types=[DT_INT32, DT_INT32, DT_INT32, DT_INT32, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]
  eval dev: perplexity 41.61, time 0s, Sun Mar  4 15:59:49 2018.
  eval test: perplexity 41.61, time 0s, Sun Mar  4 15:59:50 2018.
  global step 311100 lr 0.001 step-time 1.10s wps 6.90K ppl 38.31 gN 25.26 bleu 0.00
  global step 311200 lr 0.001 step-time 1.11s wps 6.90K ppl 37.57 gN 12.39 bleu 0.00
  global step 311300 lr 0.001 step-time 1.12s wps 6.89K ppl 37.92 gN 17.00 bleu 0.00
  global step 311400 lr 0.001 step-time 1.11s wps 6.93K ppl 37.97 gN 22.01 bleu 0.00
  global step 311500 lr 0.001 step-time 1.12s wps 6.83K ppl 37.70 gN 15.15 bleu 0.00
  global step 311600 lr 0.001 step-time 1.12s wps 6.92K ppl 37.76 gN 18.56 bleu 0.00
  global step 311700 lr 0.001 step-time 1.11s wps 6.89K ppl 38.05 gN 22.62 bleu 0.00
  global step 311800 lr 0.001 step-time 1.13s wps 6.91K ppl 37.41 gN 12.24 bleu 0.00
  global step 311900 lr 0.001 step-time 1.10s wps 6.90K ppl 38.28 gN 19.09 bleu 0.00
  global step 312000 lr 0.001 step-time 1.10s wps 6.95K ppl 37.82 gN 19.45 bleu 0.00
# Save eval, global step 312000

I have used common settings.py in both the instances:

preprocessing = {

    # Number of samples to save in training data set
    # -1 means all available in source data set
    'samples': -1,

    # Vocab max size
    'vocab_size': 15000,

    # Whether to use joined (common) vocab for both source and destination
    # (should work well with BPE/WPM-like tokenization for our chatbot - english-english translation)
    'joined_vocab': True,

    # Whether to use BPE/WPM-like tokenization, or standard one
    'use_bpe': True,

    # Whether to use:
    # - embedded detokenizer (increases number of vocab tokens, but is more accurate)
    # - external/rule-based detokenizer (based of a bunch of rules and regular expressions -
    #   doesn't increase number of tokens in vocab, but it's hard to make a rule for every case)
    # Note, that embedded detokenizer is forced to True while using BPE-like tokenizer
    'embedded_detokenizer': True,

    # Test sets' max size
    'test_size': 100,

    ## You don't need to change anything below (internal settings)

    # Source (raw) data folder
    'source_folder': source_dir,

    # Place to save preprocessed and tokenized training set
    'train_folder': train_dir,

    # File with protected phrases for standard tokenizer
    'protected_phrases_standard_file': os.path.join(package_path, 'setup/protected_phrases_standard.txt'),

    # File with protected phrases for BPE/WPM-like tokenizer
    'protected_phrases_bpe_file': os.path.join(package_path, 'setup/protected_phrases_bpe.txt'),

    # File with blacklisted answers
    'answers_blacklist_file': os.path.join(package_path, 'setup/answers_blacklist.txt'),

    # File with detokenizer rules
    'answers_detokenize_file': os.path.join(package_path, 'setup/answers_detokenize.txt'),

    # File with replace rules for answers
    'answers_replace_file': os.path.join(package_path, 'setup/answers_replace.txt'),

    # File with blacklisted answers
    'vocab_blacklist_file': os.path.join(package_path, 'setup/vocab_blacklist.txt'),

    # File with replace rules for vocab
    'vocab_replace_file': os.path.join(package_path, 'setup/vocab_replace.txt'),

    # Number of processes to be spawned during tokenization (leave None for os.cpu_count())
    'cpu_count': None,
}

# hparams
hparams = {
    'attention': 'scaled_luong',
    'num_train_steps': 10000000,
    'num_layers': 2,
    'num_units': 512,
#    'batch_size': 128,
#    'override_loaded_hparams': True,
#    'decay_scheme': 'luong234'
#    'residual': True,
    'optimizer': 'adam',
    'encoder_type': 'bi',
    'learning_rate':0.001,
    'beam_width': 20,
    'length_penalty_weight': 1.0,
    'num_translations_per_input': 20,
#    'num_keep_ckpts': 5,

    ## You don't need to change anything below (internal settings)
    'src': 'from',
    'tgt': 'to',
    'vocab_prefix': os.path.join(train_dir, "vocab"),
    'train_prefix': os.path.join(train_dir, "train"),
    'dev_prefix': os.path.join(train_dir, "tst2012"),
    'test_prefix': os.path.join(train_dir, "tst2013"),
    'out_dir': out_dir,
    'share_vocab': preprocessing['joined_vocab'],
}

######## DO NOT TOUCH ANYTHING BELOW ########

if preprocessing['use_bpe']:
    preprocessing['embedded_detokenizer'] = True
    hparams['subword_option'] = 'spm'

preprocessing['protected_phrases_file'] = preprocessing['protected_phrases_bpe_file'] if preprocessing['use_bpe'] else preprocessing['protected_phrases_standard_file']

if preprocessing['use_bpe']:
    hparams['vocab_prefix'] += '.bpe'
    hparams['train_prefix'] += '.bpe'
    hparams['dev_prefix'] += '.bpe'
    hparams['test_prefix'] += '.bpe'

if preprocessing['joined_vocab']:
    hparams['share_vocab'] = True

**Comment by daniel-kukiela** [comment]
Take number of phrases in your corpus (from train.from or train.to file) and divide it by batch size (default - 128). You'll get number of steps. That number of steps will be one epoch. You should wait probably one epoch, than lower learning rate and train for another epoch. From that point it really depends if you should lower learning rate once again and train for another epoch, just check your chatbot.

**Comment by xchgaur** [comment]
Thanks. Considering the default train.to/train.from files that come along with nat-chatbot git repo it should not have taken this long.
Then I check on the vm to look for GPU stats, I see only one GPU being actively used: rest 3 Gpus show 0% utilization. Is there a way I can force the load to be shared among all the available GPUs?

gaurav@chatbot2:/proc/driver/nvidia/gpus$ nvidia-smi
Mon Mar  5 11:37:52 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.111                Driver Version: 384.111                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 000014A1:00:00.0 Off |                    0 |
| N/A   42C    P8    60W / 149W |   4583MiB / 11439MiB |     93%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           Off  | 0000BBDC:00:00.0 Off |                    0 |
| N/A   49C    P0    74W / 149W |    137MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           Off  | 0000DB45:00:00.0 Off |                    0 |
| N/A   67C    P0    60W / 149W |    137MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           Off  | 0000F887:00:00.0 Off |                    0 |
| N/A   61C    P0    58W / 149W |    137MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      8520      C   python                                       363MiB |
|    0     61548      C   python                                      4205MiB |
|    1      8520      C   python                                        61MiB |
|    1     61548      C   python                                        61MiB |
|    2      8520      C   python                                        61MiB |
|    2     61548      C   python                                        61MiB |
|    3      8520      C   python                                        61MiB |
|    3     61548      C   python                                        61MiB |
+-----------------------------------------------------------------------------+
gaurav@chatbot2:/proc/driver/nvidia/gpus$ ps -eaf | grep 61548
gaurav    61548  24304 99 Feb28 pts/0    13-15:49:06 python train.py
gaurav   117685   2110  0 11:39 pts/1    00:00:00 grep --color=auto 61548
gaurav@chatbot2:/proc/driver/nvidia/gpus$ 
gaurav@chatbot2:/proc/driver/nvidia/gpus$ 
gaurav@chatbot2:/proc/driver/nvidia/gpus$ ls
14a1:00:00.0  bbdc:00:00.0  db45:00:00.0  f887:00:00.0
gaurav@chatbot2:/proc/driver/nvidia/gpus$ 
gaurav@chatbot2:/proc/driver/nvidia/gpus$ 
gaurav@chatbot2:/proc/driver/nvidia/gpus$ cat 14a1\:00\:00.0/information 
Model: 		 Tesla K80
IRQ:   		 24
GPU UUID: 	 GPU-7e0049e8-0d61-65e8-06ab-25b446a4b465
Video BIOS: 	 80.21.1f.00.06
Bus Type: 	 PCIe
DMA Size: 	 40 bits
DMA Mask: 	 0xffffffffff
Bus Location: 	 14a1:00:00.0
Device Minor: 	 3
gaurav@chatbot2:/proc/driver/nvidia/gpus$ cat bbdc\:00\:00.0/information 
Model: 		 Tesla K80
IRQ:   		 25
GPU UUID: 	 GPU-b7daa9eb-1684-a116-b150-154475006eeb
Video BIOS: 	 80.21.1f.00.06
Bus Type: 	 PCIe
DMA Size: 	 40 bits
DMA Mask: 	 0xffffffffff
Bus Location: 	 bbdc:00:00.0
Device Minor: 	 0
gaurav@chatbot2:/proc/driver/nvidia/gpus$ cat db45\:00\:00.0/information 
Model: 		 Tesla K80
IRQ:   		 26
GPU UUID: 	 GPU-0c70821d-d128-5b6e-5ae3-cbc027d6a171
Video BIOS: 	 80.21.1f.00.05
Bus Type: 	 PCIe
DMA Size: 	 40 bits
DMA Mask: 	 0xffffffffff
Bus Location: 	 db45:00:00.0
Device Minor: 	 1
gaurav@chatbot2:/proc/driver/nvidia/gpus$ cat f887\:00\:00.0/information 
Model: 		 Tesla K80
IRQ:   		 27
GPU UUID: 	 GPU-97a2d67b-1ba9-16e0-f845-ec20f5424ae1
Video BIOS: 	 80.21.1f.00.05
Bus Type: 	 PCIe
DMA Size: 	 40 bits
DMA Mask: 	 0xffffffffff
Bus Location: 	 f887:00:00.0
Device Minor: 	 2
gaurav@chatbot2:/proc/driver/nvidia/gpus$

**Comment by daniel-kukiela** [comment]
Add `num_gpus` params in `hparams` section in config file and set it to the number of GPUs you want to use. But according to this: https://github.com/tensorflow/nmt you won't gain much speed.
